### ISSUE TITLE
Async generateMetadata hangs the app without any visible sign of loading 

### ISSUE THREAD

**Issue by wiwo-dev** [vercel/next.js#55524]
### Link to the code that reproduces this issue or a replay of the bug

https://github.com/wiwo-dev/generatemetadata-loading-issue
https://codesandbox.io/p/github/wiwo-dev/generatemetadata-loading-issue/main 

### To Reproduce

1. Run the application using the command: `yarn run dev`.

2. Navigate to the root path (`/`). Here, you'll find two lists of "users" retrieved from jsonplaceholder. The list at the top has the problem as it uses generateMetadata. The one at the bottom doesn't use generateMetadata.

3. The top list contains links to `/user/[userId]`, and these links have a `generateMetadata` function that fetches data. Upon clicking these links, for the initial 3 seconds, there is no visible loading indication, making users believe that something might be wrong with the link.

4. In contrast, if you click on any of the links in the bottom list, which leads to `/user-no-metadata/[userId]` that doesn't use ```generateMetadata```, you'll notice that a Loading state is triggered by Suspense, providing a clear loading indication.

### Current vs. Expected behavior

**Current behaviour**
When a user navigates to a page containing a `generateMetadata` function that fetches data from a slowly working API, there's a problem. The application appears to hang, providing no indication that it's actively loading. Notably, the `fetch` operation within `generateMetadata` doesn't trigger React's Suspense, and the browser also fails to display any loading indicators.

To simulate a slow API response, a `setTimeout` is intentionally used within the `getUsers` function.

I encountered this issue while working on a project with a tens of thousands number of subpages, and it's highly likely that a given subpage will be opened for the first time and only once, with new entries continually being added.

**Expected behaviour** 
It should trigger the Suspense or show any kind of loading indicator in the browser so the user knows that something is happening under the hood and wait for the page to be displayed.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:41 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.7.0
      npm: 8.15.0
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant Packages:
      next: 13.4.20-canary.36
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

Metadata (metadata, generateMetadata, next/head)

### Additional context

Is there a way to display a loading state or any form of indication that something is loading while still utilising the `generateMetadata` function to fetch metadata information from a slow API?

**Comment by jetaggart** [comment]
We're also seeing this issue, which is a pretty big issue with for us. We have slow endpoints that generate our metadata and right now it feels like a tradeoff between bad SEO and using this feature. Is there a workaround to get metadata on the page without blocking or getting a loading ui/suspense boundary for page navigation?

I don't know if it helps but it seems the rsc is called twice in a row. If we have generateMetadata on, the first request takes as long as the metadata does to return, which doesn't seem to be caught by any suspense/loading ui boundary. If we take out the generateMetadata, both requests are still fired, but the first one is nearly instant with the second one taking some time but that is correctly picked up by the suspense loading boundary.

**Comment by leon-marzahn** [comment]
Same issue, couldn't generateMetadata be called compile time as well with a revalidation?

**Comment by Evi-2003** [comment]
Is there any update on this?

**Comment by borispoehland** [comment]
What helps a little is using `unstable_cache` to cache all subsequent requests via ISR:

```
import { unstable_cache as cache } from 'next/cache'

const getData = cache(
  (id) => fetch(...),
  ['getData'],
  { revalidate: 300 }
)

export async function generateMetadata() {
  const data = await getData();
}
```

Now it will call block with getData on the first request (still bad), from then on always return cached data and revalidate in the background.

To not let it be blocking on the first request, you'd need a way of pre-running the getData on build time. Currently everything inside `generateMetadata` does not get called on build time

**Comment by eijil** [comment]
same !!!

**Comment by eijil** [comment]
This should be a very commonly used feature, and it is surprising that the official team has not provided any fixes.

**Comment by lpwprojects** [comment]
Looks like this issue is the same as #45418

**Comment by wdire** [comment]
Not working yet. Creating dynamic title for page shouldn't be this much of a problem. Its annoying to not having next/head, forcing to use this generateMetadata + next.js fetch(if not using next.js fetch-cached both on generateMetadata and app page you would be just fetching same data twice) but still not fixing common use case bug for half a year.

**Comment by ArianHamdi** [comment]
Same issue here.
At least I expect `generateMetadata` to trigger suspense, and also I would like to bypass calling `generateMetadata` on client-side navigation, show the page instantly, and update metadata in the background.

**Comment by jaakkohurtta** [comment]
Sad to see this has not gotten any traction from the Next team as the issue is months old already. This is very lousy UX and is making me seriously consider moving my app back to pages.

**Comment by Matt-roki** [comment]
This is now a blocker for a production release on my app. Below are two screenshots of loading times, the first is excluding generateMetadata from the page file. The second is with generateMetadata resulting in a **4 SECOND SLOWER** page load. 
<img width="577" alt="Screenshot 2024-05-06 at 16 27 51" src="https://github.com/vercel/next.js/assets/86490326/9f5130a6-767d-436b-9142-3b069509ab90">
<img width="577" alt="Screenshot 2024-05-06 at 16 29 23" src="https://github.com/vercel/next.js/assets/86490326/54d5965c-1cc6-45c0-8fed-42b28678c808">

I have already cached the request with this function 
`const getCachedData = cache(async (params) => {
  const { locale, slug, searchParams } = params;
  const preview = false;
  const domainVersion = process.env.DOMAIN_VERSION;
  return await fetchCommonData({ locale, preview, slug, domainVersion, searchParams });
});`

It is then reused in the same page file for getting the pageData and the metadata. 

Going to have to figure out a way to load the metadata after the first request is done...

**Comment by steve-marmalade** [comment]
> Same issue here. At least I expect `generateMetadata` to trigger suspense, and also I would like to bypass calling `generateMetadata` on client-side navigation, show the page instantly, and update metadata in the background.

I do hope this issue gets attention, as @ArianHamdi's comment is exactly what I hoped/expected the functionality to be.

@Matt-roki , in the meantime, I was able to implement a work-around inspired by https://github.com/vercel/next.js/issues/45418#issuecomment-1914076788 that has seemed to resolve our issues.

**Comment by popovidis** [comment]
Same issue. Metadata taking ~6 seconds.

**Comment by maxmdev** [comment]
Any updates?

**Comment by RuslanAsadov** [comment]
+, same proble,

**Comment by roman-veryovkin** [comment]
Upvote the issue as well. Waiting for the solution out of the box, please)

**Comment by Matt-roki** [comment]
Okay the best solution I have found for this is to remove the async part of generateMetadata. 

Now for me all of my seo comes from a seo and that wouldnt work without async. 

I now have a prebuild file which runs and generates a seo.json file that I can read my metadata from 

`
"scripts": {
...
    "build": "node ./utils/generateSEO.js && next build", //this here runs a script to get all my seo data and sort it accordingly.
...  
},
`

Simplified code below. 

`
export function generateMetadata({ params: { fullSlug } }) {
  try{
...
  const seoData = jsonSEO.find(v => v.locale === locale && v.slug === findSlug) || {};
`

This fixes most issues the only downside is when a seo change is made a build is needed.

**Comment by feedthejim** [comment]
Hey, I wanted to give a few updates on this.

-  like @borispoehland, if the data is cache-able you should use unstable_cache to cache it
-  however that still will cause the first hit to be slow, since it won't be cached, this is a known limitation. When we release PPR, you'll be able to statically pregenerate the metadata and combine it with a dynamic/ssr page.
- if the data truly can't be cached, we're also planning on making metadata streamable, which means that we would not block on it during render, allowing you to serve the response asap and potentially the metadata 6s+ after, like some cases in this thread.

**Comment by steve-marmalade** [comment]
Thank you for the update @feedthejim .

While those enhancements seem worthy, they also don't seem like quick wins. Did you consider a solution similar to what @ArianHamdi wrote above (at least in the short-term):
> I would like to bypass calling generateMetadata on client-side navigation, show the page instantly, and update metadata in the background.

I think that would go a long way to alleviating this issue, as we would be paying the cost on first-load only, whereas now it's per-page.

**Comment by dannytlake** [comment]
+1 for bypassing the blocker waiting for async generateMetadata to resolve during client side navigation.

**Comment by yarinsa** [comment]
+1 here. Would have been nice to have an "optimistic" behavior.

**Comment by pdelfan** [comment]
> Hey, I wanted to give a few updates on this.
> 
> * like @borispoehland, if the data is cache-able you should use unstable_cache to cache it
> * however that still will cause the first hit to be slow, since it won't be cached, this is a known limitation. When we release PPR, you'll be able to statically pregenerate the metadata and combine it with a dynamic/ssr page.
> * if the data truly can't be cached, we're also planning on making metadata streamable, which means that we would not block on it during render, allowing you to serve the response asap and potentially the metadata 6s+ after, like some cases in this thread.

Thanks for this. Is there any update on streamable metadata?

**Comment by cjcheshire** [comment]
If it does become streamable will search engines accept this?

**Comment by ebidrey** [comment]
The problem is very clear in the following code snippet using app router.

1) Page got stuck for 2 seconds in loading state while generateMetadata and TestContent are running in parallel
2) When generateMetadata finish, <Suspense> start running for 2 seconds waiting TestContent to finish.
3) The page renders.

The first point is the problem, because <Suspense> must start running from the beginning and the full page with its metadata should render together.

`
import { Suspense } from "react"

export async function generateMetadata() {
    await new Promise(resolve => setTimeout(resolve, 2000))
    return {
        title: 'Test',
        description: 'Test',
    }
}

async function TestContent() {
    await new Promise(resolve => setTimeout(resolve, 4000))
    return  <h1>Test</h1>
}

export default async function Test() {
    return (
        <Suspense fallback={<div>loading</div>}>
            <TestContent />
        </Suspense>
    )
}
`

Neither latest nor canary are working as expected. With this issue is impossible to create an elegant, efficient and SEO friendly page with app router. Please help.

**Comment by flexdinesh** [comment]
I'm working on a pretty big next.js app that's on pages router. Our pages would look somewhat like this:

```
my-app/
├─ src/
│  ├─ pages/
│  │  ├─ foo/
│  │  │  ├─ bar/
│  │  │  │  ├─ index.tsx (ssr with `getServerSideProps`)
│  │  │  ├─ index.tsx (ssr with `getServerSideProps`)
```

Our biggest pain with pages router is that `getServerSideProps` of `/foo/bar` would run every time when the user navigates from `/foo` page to `/bar` even though `bar` page shares a lot of data with `foo` page. This resulted in a sluggish navigation experience where the user will receive no UX feedback for ~500ms (that's how long our roundtrip time takes). However, pages router had `routeChangeStart` and `routeChangeComplete` event listeners which we could use to know if the user requested a client side navigation. We used that to show a loader to the user in the `/foo` page before `/foo/bar` page loaded. It was not ideal, as ideal would be able to show the loader in `/foo/bar` page but it was still better than not showing the user any feedback at all when we're waiting for the server response to continue with the navigation.

We were super excited when app router was announced with persisted layouts. We waited two years to give it some time for app router to be stable. We recently did an app router migration spike and we found that navigation from `/foo` to `/foo/bar` was super fast because of two reasons:
1. We can create a `layout.tsx` file for `/foo` that would persist when we navigate to `/foo/bar` so the bar page only has to load data necessary for `/bar`
2. Because of Suspense, we were able to take the user to the `/foo/bar` page as soon as the user clicked/pressed the link to the page and show a loader while the data was loaded for the `bar` page.

But when we experimented with app router we only tested the layouts and pages, we did not test metadata as the docs did not mention anything related to `generateMetadata ` function blocking client side navigations. It only mentions that the initial load will be blocked before the data can be streamed. It does not mention anything about suspense loader getting blocked during client side transitions.

> On the initial load, streaming is blocked until generateMetadata has fully resolved, including any content from loading.js

We're working on a very big app. Our team spent a few weeks migrating some of our pages from pages router to app router. With app router, our pages look somewhat like this:

```
my-app/
├─ src/
│  ├─ app/
│  │  ├─ foo/
│  │  │  ├─ bar/
│  │  │  │  ├─ page.tsx (has generateMetadata() and generateViewport())
│  │  │  ├─ page.tsx (has generateMetadata() and generateViewport())
│  │  │  ├─ layout.tsx
```

Things were looking good until we added the `generateMetadata` function. After adding the function to `bar` page the user receives no feedback at all until the server response comes back after navigation. The user is now just waiting there not knowing what happened after clicking/pressing on a link for approxmiately ~400ms when their connection is fast, when it's slow, the user is waiting without any feeback for ~2s. This is terrible for UX. Unlike pages router, we don't have something like `routeChangeStart` and `routeChangeComplete` in the app router, so we can't show any loading UX even on the current page. Now we're caught between the pages router and the app router world with little to no hope. 

@feedthejim none of the options you suggested would work for us. It's a self-hosted app. Our data is very fast and cached in a different layer and not in the render layer for freshness reasons. We can't use `unstable_cache`, PPR or `use cache`. We don't want to bother setting up our infra with streaming either. We just want to be able to server render a fully dynamic page with persisted layouts and not have `generateMetadata` and `generateViewport` functions block the Suspense loader. Our UX was a lot better in pages router even though we were loading more data than necessary during client-side navigations to `/foo/bar` page.

**Comment by ebidrey** [comment]
> I'm working on a pretty big next.js app that's on pages router. Our pages would look somewhat like this:
> 
> 
> 
> ```
> 
> my-app/
> 
> ├─ src/
> 
> │  ├─ pages/
> 
> │  │  ├─ foo/
> 
> │  │  │  ├─ bar/
> 
> │  │  │  │  ├─ index.tsx (ssr with `getServerSideProps`)
> 
> │  │  │  ├─ index.tsx (ssr with `getServerSideProps`)
> 
> ```
> 
> 
> 
> Our biggest pain with pages router is that `getServerSideProps` of `/foo/bar` would run every time when the user navigates from `/foo` page to `/bar` even though `bar` page shares a lot of data with `foo` page. This resulted in a sluggish navigation experience where the user will receive no UX feedback for ~500ms (that's how long our roundtrip time takes). However, pages router had `routeChangeStart` and `routeChangeComplete` event listeners which we could use to know if the user requested a client side navigation. We used that to show a loader to the user in the `/foo` page before `/foo/bar` page loaded. It was not ideal, as ideal would be able to show the loader in `/foo/bar` page but it was still better than not showing the user any feedback at all when we're waiting for the server response to continue with the navigation.
> 
> 
> 
> We were super excited when app router was announced with persisted layouts. We waited two years to give it some time for app router to be stable. We recently did an app router migration spike and we found that navigation from `/foo` to `/foo/bar` was super fast because of two reasons:
> 
> 1. We can create a `layout.tsx` file for `/foo` that would persist when we navigate to `/foo/bar` so the bar page only has to load data necessary to render bar
> 
> 2. Because of Suspense, we were able to take the user to the `/foo/bar` page as soon as the user clicked/pressed the link to the page and show a loader while the data was loaded for the `bar` page.
> 
> 
> 
> But when we experimented with app router we only tested the layouts and pages, we did not test metadata as the docs did not mention anything related to `generateMetadata` function blocking client side navigations. It only mentions that the initial load will be blocked before the data can be streamed. It does not mention anything about suspense loader getting blocked during client side transitions.
> 
> 
> 
> > On the initial load, streaming is blocked until generateMetadata has fully resolved, including any content from loading.js
> 
> 
> 
> We're working on a very big app. Our team spent a few weeks migrating some of our pages from pages router to app router. With app router, our pages look somewhat like this:
> 
> 
> 
> ```
> 
> my-app/
> 
> ├─ src/
> 
> │  ├─ app/
> 
> │  │  ├─ foo/
> 
> │  │  │  ├─ bar/
> 
> │  │  │  │  ├─ page.tsx (has generateMetadata() and generateViewport())
> 
> │  │  │  ├─ page.tsx (has generateMetadata() and generateViewport())
> 
> │  │  │  ├─ layout.tsx
> 
> ```
> 
> 
> 
> Things were looking good until we added the `generateMetadata` function. After adding the function to `bar` page the user receives no feedback at all until the server response comes back after navigation. The user is now just waiting there not knowing what happened after clicking/pressing on a link for approxmiately ~400ms when their connection is fast, when it's slow, the user is waiting without any feeback for ~2s. This is terrible for UX. Unlike pages router, we don't have something like `routeChangeStart` and `routeChangeComplete` in the app router, so we can't show any loading UX even on the current page. Now we're caught between the pages router and the app router world with little to no hope. 
> 
> 
> 
> @feedthejim none of the options you suggested would work for us. It's a self-hosted app. Our data is very fast and cached in a different layer and not in the render layer for freshness reasons. We can't use `unstable_cache`, PPR or `use cache`. We don't want to bother setting up our infra with streaming either. We just want to be able to server render a fully dynamic page with persisted layouts and not have `generateMetadata` and `generateViewport` functions block the Suspense loader. Our UX was a lot better in pages router even though we were loading more data than necessary during client-side navigations to `/foo/bar` page.

Exactly the same issue here, caught between page and app router.

**Comment by philwolstenholme** [comment]
👀 https://x.com/stewiemcstews/status/1863680862308667716

**Comment by huozhi** [comment]
👋  Hi everyone!  We heard your requests. 


We're introducing **Streaming Metadata** as the enhancement of the recent canary releases to improve metadata rendering performance. 

### Introduction

This feature is designed to improve page loading time for users, especially when dealing with slow metadata generation. With Streaming Metadata enabled:
* The `generateMetadata` returned metadata is treated as suspenseful data, allowing the page content to render immediately.
* Metadata is then injected asynchronously into the page on the client side once it’s resolved.

Since metadata is critical for SEO, we’ve ensured that bots continue to receive fully rendered metadata in the HTML. For human users, who primarily care about page content, metadata can be added later without impacting their experience.

### Usage
The following new options in `next.config.js` are for configuring Streaming Metadata:
- `experimental.streamingMetadata` **(boolean)**: Enables the Streaming Metadata feature _(disabled by default)_
- `experimental.htmlLimitedBots` **optional (RegExp)**: Specifies a regex to identify bot user agents that should opt out of the streaming metadata feature and continue receiving fully blocking metadata in the HTML. _(We have a list of common bots which have limited ability to handle metadata rendering)_

By default, only the Google bots which can act like headless browsers will receive streaming metadata when you enable the feature. Other bots will receive blocking metadata like the ones configured in `htmlLimitedBots`. If you noticed any bots also need to consume blocking metadata, please let us know to add it into default list.

If you don't need to customize the bots user agents for the blocking metadata and default config is enough for you, then only need to have one flag enabled in your `next.config.js`

```js
module.exports = {
  experimental: {
    streamingMetadata: true,
  },
}
```

### Notes
We’re still working on optimizations for PPR mode. However, if you’re not using PPR, you can try out this feature now to accelerate pages with slow `generateMetadata()` calls.

We’d love to hear your thoughts! Please feel free to share any comments, suggestions, or concerns in this thread. We'd love to hear your feedbacks! Thank you! 🚀

**Comment by ssotomayor** [comment]
@huozhi 
> 👋 Hi everyone! We heard your requests.
> 
> We're introducing **Streaming Metadata** as the enhancement of the recent canary releases to improve metadata rendering performance.
> 
> ### Introduction
> This feature is designed to improve page loading time for users, especially when dealing with slow metadata generation. With Streaming Metadata enabled:
> 
> * The `generateMetadata` returned metadata is treated as suspenseful data, allowing the page content to render immediately.
> * Metadata is then injected asynchronously into the page on the client side once it’s resolved.
> 
> Since metadata is critical for SEO, we’ve ensured that bots continue to receive fully rendered metadata in the HTML. For human users, who primarily care about page content, metadata can be added later without impacting their experience.
> 
> ### Usage
> The following new options in `next.config.js` are for configuring Streaming Metadata:
> 
> * `experimental.streamingMetadata` **(boolean)**: Enables the Streaming Metadata feature _(disabled by default)_
> * `experimental.htmlLimitedBots` **optional (RegExp)**: Specifies a regex to identify bot user agents that should opt out of the streaming metadata feature and continue receiving fully blocking metadata in the HTML. _(We have a list of common bots which have limited ability to handle metadata rendering)_
> 
> By default, only the Google bots which can act like headless browsers will receive streaming metadata when you enable the feature. Other bots will receive blocking metadata like the ones configured in `htmlLimitedBots`. If you noticed any bots also need to consume blocking metadata, please let us know to add it into default list.
> 
> If you don't need to customize the bots user agents for the blocking metadata and default config is enough for you, then only need to have one flag enabled in your `next.config.js`
> 
> module.exports = {
>   experimental: {
>     streamingMetadata: true,
>   },
> }
> ### Notes
> We’re still working on optimizations for PPR mode. However, if you’re not using PPR, you can try out this feature now to accelerate pages with slow `generateMetadata()` calls.
> 
> We’d love to hear your thoughts! Please feel free to share any comments, suggestions, or concerns in this thread. We'd love to hear your feedbacks! Thank you! 🚀

Just tested this on latest canary, it's amazing when running with `next dev` with turbopack, exactly what's expected, however when i make a production build i don't have the same behaviour, the navigation becomes slow yet again on first navigation.

**Comment by huozhi** [comment]
@ssotomayor Thanks for the feedback, can you share a bit more about user case? A reproduction would be the best, I can help look into it.

**Comment by pondorasti** [comment]
> We're introducing Streaming Metadata as the enhancement of the recent canary releases to improve metadata rendering performance.

yesss, this makes me so happy! e-comm style apps with slow metadata has been one of my biggest gripes with Next.js as it absolutely tanks INP for link navigation. I just migrated the entirety of [emojis.com](https://www.emojis.com/) to the latest canary with `streamingMetadata` and results are positive!

**Comment by JamesDonnelly** [comment]
@huozhi is there any reason you can think that the canary build would break both `metadata` and `generateMetadata` completely for social media previews?

On a fresh Next.js install using the latest canary build everything works fine, but on an existing project as soon as I bump Next.js from 15.0.1 to 15.2.0-canary.19 even simple `metadata` exports which were working before fail to generate. This happens regardless of whether I enable `streamingMetadata` or not.

```ts
export const metadata: Metadata = {
  openGraph: {
    description: 'Test',
    title: 'Test',
  },
}
```

Unfortunately this project is private and is not mine to distribute, but I was unable to reproduce the problem. I realise this is an open ended question, but I'm wondering if this is something you or anyone else may have experienced with this already?

I'm seeing 200 responses everywhere in Vercel's logs, and on more complex `generateMetadata` calls I'm seeing 200 responses returned from the external APIs those call, but resources like Discord and Facebook never see the OG tags.

The project itself is not behind Vercel authentication or anything that would otherwise get in the way.

![Image](https://github.com/user-attachments/assets/8f84ae66-7376-481e-9036-9d8dad4373b2)

![Image](https://github.com/user-attachments/assets/8e8a8278-8e23-47f4-9227-f18bf7984b84)

**Comment by MagicalTurtle** [comment]
> 👋 Hi everyone! We heard your requests.
> 
> We're introducing **Streaming Metadata** as the enhancement of the recent canary releases to improve metadata rendering performance.
> 
> ### Introduction
> This feature is designed to improve page loading time for users, especially when dealing with slow metadata generation. With Streaming Metadata enabled:
> 
> * The `generateMetadata` returned metadata is treated as suspenseful data, allowing the page content to render immediately.
> * Metadata is then injected asynchronously into the page on the client side once it’s resolved.
> 
> Since metadata is critical for SEO, we’ve ensured that bots continue to receive fully rendered metadata in the HTML. For human users, who primarily care about page content, metadata can be added later without impacting their experience.
> 
> ### Usage
> The following new options in `next.config.js` are for configuring Streaming Metadata:
> 
> * `experimental.streamingMetadata` **(boolean)**: Enables the Streaming Metadata feature _(disabled by default)_
> * `experimental.htmlLimitedBots` **optional (RegExp)**: Specifies a regex to identify bot user agents that should opt out of the streaming metadata feature and continue receiving fully blocking metadata in the HTML. _(We have a list of common bots which have limited ability to handle metadata rendering)_
> 
> By default, only the Google bots which can act like headless browsers will receive streaming metadata when you enable the feature. Other bots will receive blocking metadata like the ones configured in `htmlLimitedBots`. If you noticed any bots also need to consume blocking metadata, please let us know to add it into default list.
> 
> If you don't need to customize the bots user agents for the blocking metadata and default config is enough for you, then only need to have one flag enabled in your `next.config.js`
> 
> module.exports = {
>   experimental: {
>     streamingMetadata: true,
>   },
> }
> ### Notes
> We’re still working on optimizations for PPR mode. However, if you’re not using PPR, you can try out this feature now to accelerate pages with slow `generateMetadata()` calls.
> 
> We’d love to hear your thoughts! Please feel free to share any comments, suggestions, or concerns in this thread. We'd love to hear your feedbacks! Thank you! 🚀

We ran into issues with metadata tags not being set correctly when a bot requests the page.

Steps to reproduce:
    - Enable setting
    - Set open graph meta tags
    - Use a valid bot (discord, twitter, etc..) and post a link so it tries to generate open graph

Expected:
    - Bot correctly displays OG embed with title, description and image

Actual:
    - Bot can't read any of the metadata tags

We checked manually as well by getting a page with metadata with the discordbot user agent (`Mozilla/5.0 (compatible; Discordbot/2.0; +https://discordapp.com)`) and we can confirm the metadata tags do not get set on `<head>`. We tried this both on a staging environment and a local environment.

We can also confirm that user agent is being passed correctly when we trigger discord to embed the link

**Comment by JamesDonnelly** [comment]
@MagicalTurtle I've been having the same problem. Switching from v15.0.1 to the latest canary build broke metadata before we even enabled `streamingMetadata`. It broke both `metadata` and `generateMetadata`, despite 200 responses being shown everywhere relevant.

I did some testing and found that authentication calls through Firebase seemed to be causing the issues, but we couldn't narrow it down further than that.

I was able to work around it by removing auth-related references from my layout file.

The `RootLayout` I have previously looked like this:

```ts
// layout.tsx
export default async function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  const user = await getUser();

  return (
    <AppRouterCacheProvider>
      <html lang="en">
        <body>
          <ThemeProvider theme={theme}>
            <UIProvider>
              <AuthContextProvider user={user}>
                <APIContextProvider>
                  <EditorContextProvider canEdit={!!user?.customClaims['index-editor']}>
                    <PageHeader />
                    {children}
                  </EditorContextProvider>
                </APIContextProvider>
              </AuthContextProvider>
            </UIProvider>
          </ThemeProvider>
        </body>
      </html>
    </AppRouterCacheProvider>
  );
}
```

Moving everything from `AuthContextProvider` down into each page component instead made it work as expected:

```ts
// layout.tsx
export default async function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <AppRouterCacheProvider>
      <html lang="en">
        <body>
          <ThemeProvider theme={theme}>
            <UIProvider>
              {children}
            </UIProvider>
          </ThemeProvider>
        </body>
      </html>
    </AppRouterCacheProvider>
  );
}
```

```
// page.tsx
export default async function Page() {
  const user = await getUser();
  
  return (
    <AuthContextProvider user={user}>
      <APIContextProvider>
        <EditorContextProvider canEdit={!!user?.customClaims['index-editor']}>
          <PageHeader />
          <PageContent>
            <Dashboard />
          </PageContent>
        </EditorContextProvider>
      </APIContextProvider>
    </AuthContextProvider>
  );
}
```

The big downside to this is having to duplicate those providers on every page. My assumption is that the current logic puts too much emphasis on what happens in the layout and doesn't skip straight to the _is this a bot and is there metadata that can be served?_ step.

**Comment by huozhi** [comment]
@MagicalTurtle @JamesDonnelly are you generating static page? if so it's probably related to a issue I'm fixing right now https://github.com/vercel/next.js/pull/75227

**Comment by JamesDonnelly** [comment]
> are you generating static page?

I have a mix of static and dynamic pages which share the same layout. Metadata was breaking everywhere.
If it helps I was seeing the metadata pulled through into the HTML locally, so the async client side change was working as expected, however the metadata wasn't making it through for bots.

**Comment by MagicalTurtle** [comment]
> [@MagicalTurtle](https://github.com/MagicalTurtle) [@JamesDonnelly](https://github.com/JamesDonnelly) are you generating static page? if so it's probably related to a issue I'm fixing right now [#75227](https://github.com/vercel/next.js/pull/75227) Since I don't have a reproduction from of your case, hard to tell what's the issue.

Hmm, i'm not sure what constitutes a static page. We're using the `standalone` output but my page does not use any dynamic apis (`cookies()`, `headers()`, etc..)

Here are my layout and home pages

```tsx
export const metadata: Metadata = {....}

export default async function RootLayout({
  children
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en" className={`${inter.variable} ${satoshi.variable} antialiased`}>
      <body>
        <div id="root">
          {children}
        </div>
      </body>
    </html>
  );
```

```tsx
export const revalidate = 3600;

export default function Home() {
  return <div>...</div>;
}

```

**Comment by MagicalTurtle** [comment]
We've updated next version to `15.2.0-canary.30` and it seems to work as expected now. Bots can get metadata correctly from the head and non-bot agents get their metadata streamed (albeit a bit slower than canary 20 but workable)

**Comment by huozhi** [comment]
Hi everyone, we've shipped the streaming metadata as the default behavior to canary release, you don't need to have that experimental configuration `streamingMetadata` anymore. If you still need to customized the html limited bots, you can directly update `htmlLimitedBots` in `next.config.js` which is a stable configuration now.

**Comment by kachkaev** [comment]
Great news, thank you @huozhi! One small source of confusion: would setting `htmlLimitedBots` override default ones? It feels a bit scary to edit this setting after reading https://nextjs.org/docs/app/api-reference/functions/generate-metadata#streaming-metadata. It’d be also great to know what the default list is and under what principles it can be updated. For example, if I find a regional bot (e.g. in a country with 50M people), would it go to the global list via a PR? Or is the shared list dedicated to global players only?

**Comment by philwolstenholme** [comment]
If the default list of HTML limited bots was exported by Next.js then we could spread it into our new list when setting `htmlLimitedBots` in our config. That way we could add bots while not losing our on any changes Next.js made to it's default list of bots.

**Comment by kachkaev** [comment]
UPD: found this in Next.js source:

https://github.com/vercel/next.js/blob/13fcfabf1beb3565d2bb191bd8d8ce4ad6d97b55/packages/next/src/shared/lib/router/utils/is-bot.ts#L5-L8

→ the list contains a bunch of regional players, but there is Telegram, for example (regional but quite big)

---

Looks like setting `htmlLimitedBots` overrides the default value (no concatenation):

https://github.com/vercel/next.js/blob/13fcfabf1beb3565d2bb191bd8d8ce4ad6d97b55/packages/next/src/server/lib/streaming-metadata.ts#L21-L24

https://github.com/vercel/next.js/blob/13fcfabf1beb3565d2bb191bd8d8ce4ad6d97b55/packages/next/src/server/config.ts#L1046-L1049

**Comment by huozhi** [comment]
Good call, I'll update the docs about the default pattern!

**Comment by sean-ahn** [comment]
This is an essential feature. Some search engines with a large user base in specific regions or business domains may not operate a headless browser when crawling.

However, overriding the default settings (current behavior) would mean losing the benefits of an actively updated list in the future.

To address this, I believe the htmlLimitedBots config should provide a feature to extend the default values rather than completely overriding them.

**Comment by huozhi** [comment]
@sean-ahn Do you have any special search engine that not covered in the default list that needs to be addressed? We can add them into the list. Ideally you don't have to override them most of the time and defaut list can cover them.

**Comment by sean-ahn** [comment]
> [@sean-ahn](https://github.com/sean-ahn) Do you have any special search engine that not covered in the default list that needs to be addressed? We can add them into the list. Ideally you don't have to override them most of the time and defaut list can cover them.

That approach is exactly what we need.
Could you add `Yeti` from **Naver** (the most popular search engine in South Korea, serving a population of >50M) to the default list?

#### User Agent Example
```
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko; compatible; Yeti/1.1; +https://naver.me/spd) Chrome/W.X.Y.Z Safari/537.36
```
or 
```
Mozilla/5.0 (compatible; Yeti/1.1; +https://naver.me/spd)
```

Also, since Naver hasn’t explicitly stated whether Yeti uses a headless browser engine, I think it seems safer to include it in the list rather than relying on assumptions.

**Comment by sean-ahn** [comment]
> [@sean-ahn](https://github.com/sean-ahn) Do you have any special search engine that not covered in the default list that needs to be addressed? We can add them into the list. Ideally you don't have to override them most of the time and defaut list can cover them.

@huozhi Sounds good. That approach is exactly what we need.
Could you add `Yeti` from **Naver** (the most popular search engine in South Korea, serving a population of >50M) to the default list?

#### User Agent Example
```
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko; compatible; Yeti/1.1; +https://naver.me/spd) Chrome/W.X.Y.Z Safari/537.36
```
or 
```
Mozilla/5.0 (compatible; Yeti/1.1; +https://naver.me/spd)
```

Also, since Naver hasn’t explicitly stated whether Yeti uses a headless browser engine, I think it seems safer to include it in the list rather than relying on assumptions.

**Comment by huozhi** [comment]
@sean-ahn I'll add that one later to the default list 🙏
